src/server/project.ts

```diff
@@ -116,8 +88,23 @@ namespace ts.server {
         private missingFilesMap: Map<FileWatcher>;
         private plugins: PluginModule[] = [];
 
-        private cachedUnresolvedImportsPerFile = new UnresolvedImportsMap();
-        private lastCachedUnresolvedImportsList: SortedReadonlyArray<string>;
+        /*@internal*/
+        /**
+         * This is map from files to unresolved imports in it
+         * Maop does not contain entries for files that do not have unresolved imports
+         * This helps in containing the set of files to invalidate
+         */
+        cachedUnresolvedImportsPerFile = createMap<ReadonlyArray<string>>();
+
+        /**
+         * This is the set that has entry to true if file doesnt contain any unresolved import
+         */
+        private filesWithNoUnresolvedImports = createMap<true>();
+
+        /*@internal*/
+        lastCachedUnresolvedImportsList: SortedReadonlyArray<string>;
+        /*@internal*/
+        hasMoreOrLessScriptInfos = false;

```

Comment:
consider renaming this to hasAddedOrRemovedFiles

Reply:
how is that different from is this the same as projectStructureVersion change?

Reply:
projectStructureVersion is actually program version = every time new program is created this version is incremented. We could create new program using existing files (without any additional or removal of script infos when we decide its safe to reuse the modules, eg. adding import in file such that imported module was already part of the program, so it would be new program but no additional files woult be there in the program)